File: Algorithm and Data Structure/Baekjoon/2000번대/02468번 안전 영역/boj2468_rebas.py
import sys
sys.setrecursionlimit(100000)

def dfs(x, y, z):
    check[x][y] = True
    for dx, dy in (-1,0), (1,0), (0,-1), (0,1):
        nx, ny = x+dx, y+dy
        if nx < 0 or nx >= n or ny < 0 or ny >= n:
            continue
        if not check[nx][ny] and a[nx][ny] > z:
            dfs(nx, ny, z)

n = int(input())
a = [list(map(int, input().split())) for _ in range(n)]
ans = 0
for k in range(max(max(a))):
    check = [[False]*n for _ in range(n)]
    cnt = 0
    for i in range(n):
        for j in range(n):
            if not check[i][j] and a[i][j] > k:
                dfs(i, j, k)
                cnt += 1
    ans = max(ans, cnt)
print(ans)

File: Algorithm and Data Structure/Baekjoon/2000번대/02400번대/02468번 안전 영역/boj2468_rebas.cc
#include <cstdio>
#include <cstring>
using namespace std;

struct cheese {
    int x, y;
};

int n, m, ans;
int a[100][100];
bool check[100][100];
const int dx[] = {-1, 0, 1, 0}, dy[] = {0, -1, 0, 1};

void dfs(int x, int y, int z) {
    check[x][y] = true;
    for (int i=0; i<4; i++) {
        int nx = x+dx[i], ny = y+dy[i];
        if (nx < 0 || nx >= n || ny < 0 || ny >= n) continue;
        if (check[nx][ny] || a[nx][ny] <= z) continue;
        dfs(nx, ny, z);
    }
}

int main() {
    scanf("%d", &n);
    for (int i=0; i<n; i++) {
        for (int j=0; j<n; j++) {
            scanf("%d", &a[i][j]);
            if (m < a[i][j]) m = a[i][j];
        }
    }
    for (int k=0; k<m; k++) {
        memset(check, false, sizeof(check));
        int cnt = 0;
        for (int i=0; i<n; i++) {
            for (int j=0; j<n; j++) {
                if (!check[i][j] && a[i][j] > k) {
                    dfs(i, j, k);
                    cnt++;
                }
            }
        }
        if (ans < cnt) ans = cnt;
    }
    printf("%d\n", ans);
    return 0;
}


